File: app/models/recipe.rb
# frozen_string_literal: true

##
# == Fields
# name:             string
# category:         ApplicationRecord::Category
# user:             ApplicationRecord::User
# preparation_time: integer
# cooking_time:     integer
# number_of_people: integer
# difficulty:       string
# price:            string
# image:            string
# preparation:      text
# user_comment:     string
# created_at:       datetime
# updated_at:       datetime
class Recipe < ApplicationRecord
  
  belongs_to :user
  belongs_to :category
  has_many :ingredients_recipes
  has_many :ingredients, through: :ingredients_recipes
  has_and_belongs_to_many :kitchenwares
  has_and_belongs_to_many :countries
  has_many :recipe_images

  validates :name, presence: true
  validates :preparation_time,
            :cooking_time,
            :number_of_people,
                allow_nil: true, numericality: { greater_than: 0 }
  enum difficulty: %i[easy regular hard]
  enum price: %i[low medium high]
  validates_presence_of :image, :preparation

end

File: frontend/src/dto/GET/recipe.ts
import type { RecipeDifficulty, RecipePrice } from '@/models/recipe'

export default interface GetRecipeDTO {
  id: number
  image: string
  name: string
  country_id: number | null
  category_id: number
  user_id: number
  preparation_time: number | null
  cooking_time: number | null
  number_of_people: number | null
  difficulty: RecipeDifficulty
  price: RecipePrice
  created_at: string
  updated_at: string
  description: string

  category: {
    id: number
    name: string
  }
  country?: {
    id: number
    image: string | null
    name: string
  } | null
  user: {
    id: number
    email: string
  }

  kitchenwares?: Array<GetRecipeKitchenwareDTO>
  ingredients?: Array<GetRecipeIngredientDTO>
  instructions?: Array<GetRecipeInstructionDTO>
}

export interface GetRecipeKitchenwareDTO {
  id: number
  name: string
  image: string | null
}

export interface GetRecipeIngredientDTO {
  id: number
  ingredient_id: number
  image: string | null
  name: string
  quantity: number | null
  measure: string | null
  comment: string | null
}

export interface GetRecipeInstructionDTO {
  id: number
  step: number
  comment: string
}
